libsplinter/src/network/connection_manager/mod.rs

```diff
@@ -0,0 +1,642 @@
+    hb_monitor: HeartbeatMonitor,
+    connection_state: Arc<Mutex<ConnectionState>>,
+    join_handle: Option<thread::JoinHandle<()>>,
+    sender: Option<Sender<CmMessage>>,
+}
+
+impl ConnectionManager {
+    pub fn new(mesh: Mesh, transport: Box<dyn Transport + Send>) -> Self {
+        let connection_state = Arc::new(Mutex::new(ConnectionState::new(mesh, transport)));
+        let hb_monitor =
+            HeartbeatMonitor::new(connection_state.clone(), DEFAULT_HEARTBEAT_INTERVAL);
+
+        Self {
+            hb_monitor,
+            connection_state,
+            join_handle: None,
+            sender: None,
+        }
+    }
+
+    pub fn start(&mut self) -> Result<Connector, ConnectionManagerError> {
+        let (sender, recv) = unbounded();
+        let connection_state = self.connection_state.clone();
+
+        let join_handle = thread::Builder::new()
+            .name("Connection Manager".into())
+            .spawn(move || {
+                let mut subscribers = HashMap::new();
+                loop {
+                    match recv.recv() {
+                        Ok(CmMessage::Shutdown) => break,
+                        Ok(CmMessage::Subscribe(id, sender)) => {
+                            subscribers.insert(id, sender);
+                        }
+                        Ok(CmMessage::UnSubscribe(ref id)) => {
+                            subscribers.remove(id);
+                        }
+                        Ok(CmMessage::Request(req)) => {
+                            let mut state = match connection_state.lock() {
+                                Ok(s) => s,
+                                Err(err) => {
+                                    notify_subscribers(
+                                        &mut subscribers,
+                                        vec![CmNotification::FatalError {
+                                            error: ConnectionManagerError::StatePoisoned,
+                                            message: format!("State has been poisoned {:?}", err),
+                                        }],
+                                    );
+                                    break;
+                                }
+                            };
+
+                            handle_request(req, &mut state);
+                        }
+                        Ok(CmMessage::HeartbeatNotifications(notifications)) => {
+                            notify_subscribers(&mut subscribers, notifications)
+                        }
+                        Err(_) => {
+                            warn!("All senders have disconnected");
+                            break;
+                        }
+                    }
+                }
+            })?;
+
+        self.hb_monitor.start(sender.clone())?;
+
+        self.join_handle = Some(join_handle);
+        self.sender = Some(sender.clone());
+
+        Ok(Connector { sender })
+    }
+
+    pub fn shutdown(self) {
```

Comment:
Who will own this object to call shutdown?

Reply:
`SplinterDaemon` would create and own the ConnectionManager, and anything that needs to talk to it would own a `Connector`

Reply:
Added the following.
`ShutdownHandle`: Send shutdown message to connection manager thread
`ConnectionManager::shutdown_handle`: Returns `Some(ShutdownHandle)` if connection manager has been started and `None` otherwise
`ConnectionManager::await_shutdown`: Joins connection manager threads
`ConnectionManager::shutdown_and_wait`: Convenience method that sends a shutdown message and joins threads